src/relay/pass/pass_manager.cc

```diff
@@ -0,0 +1,569 @@
+            << " with opt level: " << pass_info.operator->()->opt_level << "\n";
+  CHECK(mod.defined());
+  std::vector<std::pair<GlobalVar, Function>> updated_funcs;
+  ModuleNode* mod_node = mod.operator->();
+  for (const auto& it : mod_node->functions) {
+    if (!SkipFunction(it.second)) {
+      auto updated_func = pass_func(it.second, pass_ctx_);
+      CHECK(updated_func.defined());
+      updated_funcs.push_back({std::move(it.first), std::move(updated_func)});
+    }
+  }
+
+  // Update the optimized functions.
+  for (const auto& it : updated_funcs) {
+    mod_node->Update(it.first, it.second);
+  }
+
+  return GetRef<Module>(mod_node);
+}
+
+void FunctionPassNode::SetContext(const PassContext& pass_ctx) {
+  pass_ctx_ = pass_ctx;
+}
+
+// TODO(zhiics) Create an enum attribute for FunctionNode
+// enum Attribute {kPrimitive, kSkipOptimization}
+bool FunctionPassNode::SkipFunction(const Function& func) const {
+  NodeRef res = FunctionGetAttr(func, "SkipOptimization");
+  const ir::IntImm* pval = res.as<ir::IntImm>();
+  return pval && pval->value != 0;
+}
+
+std::vector<std::string> FunctionPassNode::Required() const {
+  std::vector<std::string> required;
+  PassInfo pass_info = GetPassInfo();
+  CHECK(pass_info.defined())
+      << "Cannot get the required passes from undefined PassInfo object."
+      << "\n";
+  for (const auto& it : pass_info.operator->()->required) {
+    const auto* str = it.as<tvm::ir::StringImm>();
+    required.push_back(str->value);
+  }
+  return required;
+}
+
+SequentialPass SequentialPassNode::make(PassInfo pass_info,
+                                        tvm::Array<Pass> passes,
+                                        tvm::Array<tvm::Expr> disabled) {
+  auto n = make_node<SequentialPassNode>();
+  n->pass_info = std::move(pass_info);
+  n->passes = std::move(passes);
+  n->disabled = std::move(disabled);
+  return SequentialPass(n);
+}
+
+// TODO(jroesch, zhiics): we currenlty only sequentially execute each pass in
+// a SequentialPass without the consideration of their orders. The phase
+// ordering problem needed to be handled in the future.
+Module SequentialPassNode::operator()(const Module& module) const {
+  Module mod = module;
+  for (const Pass& pass : passes) {
+    CHECK(pass.defined()) << "Found undefined pass for optimization.";
+    const auto* pn = pass.operator->();
+    mod = (*pn)(mod);
+  }
+  return mod;
+}
+
+std::vector<std::string> SequentialPassNode::Required() const {
+  std::vector<std::string> required;
+  PassInfo pass_info = GetPassInfo();
+  CHECK(pass_info.defined())
+      << "Cannot get the required passes from undefined PassInfo object."
+      << "\n";
+  for (const auto& it : pass_info.operator->()->required) {
+    const auto* str = it.as<tvm::ir::StringImm>();
+    required.push_back(str->value);
+  }
+  return required;
+}
+
+void SequentialPassNode::ResolveDependency(const Module& mod) {
+  // TODO(zhiics) Implement it.
```

Comment:
How about throwing an error or removing the function for now?

Reply:
done